### PR TITLE
Solves Issue #32750 - torch.prod now works fine with FP16 Input Tensor and FP32 Output Tensor

### PR DESCRIPTION
**Pull request by krshrimali**
This PR solves Issue #32750. Changes function `prod_kernel_impl` to use `acc_t` argument instead of `scalar_t` (which caused the garbage output for FP16 input and FP32 output tensor type).

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit dc0e6f35:


**None of the build failures appear to be your fault**.

* **1/1** recognized as flaky :snowflake:
    * Re-run these jobs?


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=dc0e6f35f2f0dc05aae73967b9bee2a8006c18bd).

### :snowflake: 1 failure recognized as flaky
The following build failures have been detected as flaky and **may not be your fault**:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4370257) pytorch_windows_test2 (1/1)
**Step:** "Test" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=92703656) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=92703656)) :snowflake:
<details>
<summary>
<code>AssertionError: tensor(1.1904) not less than or equal to 0.01 :</code>
</summary>

```
---------------------------------------------------------------------- 
Traceback (most recent call last): 
  File "C:\Users\circleci\project\build\win_tmp\build\torch\testing\_internal\common_device_type.py", line 197, in instantiated_test 
    result = test(self, device_arg, dtype) 
  File "test_torch.py", line 14549, in fn 
    self.assertEqual(cpu_result, device_result, prec=precision) 
  File "C:\Users\circleci\project\build\win_tmp\build\torch\testing\_internal\common_utils.py", line 862, in assertEqual 
    assertTensorsEqual(x, y) 
  File "C:\Users\circleci\project\build\win_tmp\build\torch\testing\_internal\common_utils.py", line 832, in assertTensorsEqual 
    self.assertLessEqual(max_err, prec, message) 
AssertionError: tensor(1.1904) not less than or equal to 0.01 :  
 
---------------------------------------------------------------------- 
Ran 2911 tests in 45.959s 
 
FAILED (failures=1, errors=741, skipped=43) 
Traceback (most recent call last): 
  File "run_test.py", line 457, in <module> 
    main() 
  File "run_test.py", line 450, in main 
    raise RuntimeError(message) 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

**Comment by yf225** [comment]
@nairbv Would you be interested in reviewing it?

File: aten/src/ATen/native/cuda/ReduceOpsKernel.cu [hunk 1]
```diff
@@ -41,7 +41,7 @@ void std_var_kernel_impl<at::Half>(TensorIterator& iter, bool unbiased, bool tak
 
 template <typename scalar_t, typename acc_t=scalar_t>
 void prod_kernel_impl(TensorIterator& iter) {
-  gpu_reduce_kernel<scalar_t, scalar_t>(iter, func_wrapper<scalar_t> ([]GPU_LAMBDA(acc_t a, acc_t b) -> acc_t {
+  gpu_reduce_kernel<scalar_t, acc_t>(iter, func_wrapper<scalar_t> ([]GPU_LAMBDA(acc_t a, acc_t b) -> acc_t {
```
Comment (nairbv) [inline aten/src/ATen/native/cuda/ReduceOpsKernel.cu]: we should have a test

Comment (krshrimali) [reply aten/src/ATen/native/cuda/ReduceOpsKernel.cu]: Sure, @nairbv. I might have to do a few more modifications to the `prod_kernel_impl` as well. And also write a unit test, & commit. Thanks!

**Comment by krshrimali** [comment]
Closing this PR, as I seemed to have messed up things while trying to re-base.

**Pull request closed without merging**